go/adbc/driver/bigquery/connection.go

```diff
@@ -0,0 +1,868 @@
+		}
+	case "ARRAY":
+		arrayType := strings.Replace(typeString[:len(dataType)], "<", "", 1)
+		rIndex := strings.LastIndex(arrayType, ">")
+		if rIndex == -1 {
+			return arrow.Field{}, adbc.Error{
+				Code: adbc.StatusInvalidData,
+				Msg:  fmt.Sprintf("Cannot parse array type `%s` for field `%s`: cannot find `>`", typeString, name),
+			}
+		}
+		arrayType = arrayType[:rIndex] + arrayType[rIndex+1:]
+		arrayFieldType, err := buildField(name, typeString, arrayType, rIndex)
+		if err != nil {
+			return arrow.Field{}, err
+		}
+		field = arrayFieldType
+	case "RECORD", "STRUCT":
+		fieldRecords := typeString[index+1:]
+		fieldRecords = fieldRecords[:len(fieldRecords)-1]
+		nestedFields := make([]arrow.Field, 0)
+		for _, record := range strings.Split(fieldRecords, ",") {
+			fieldRecord := strings.TrimSpace(record)
+			recordParts := strings.SplitN(fieldRecord, " ", 2)
+			if len(recordParts) != 2 {
+				return arrow.Field{}, adbc.Error{
+					Code: adbc.StatusInvalidData,
+					Msg:  fmt.Sprintf("invalid field record `%s` for type `%s`", fieldRecord, dataType),
+				}
+			}
+			fieldName := recordParts[0]
+			fieldType := recordParts[1]
+			nestedField, err := buildSchemaField(fieldName, fieldType)
+			if err != nil {
+				return arrow.Field{}, err
+			}
+			nestedFields = append(nestedFields, nestedField)
+		}
+		structType := arrow.StructOf(nestedFields...)
+		if structType == nil {
+			return arrow.Field{}, adbc.Error{
+				Code: adbc.StatusInvalidArgument,
+				Msg:  fmt.Sprintf("Cannot create a struct schema for record `%s`", fieldRecords),
+			}
+		}
+		field.Type = structType
+	default:
+		return arrow.Field{}, adbc.Error{
+			Code: adbc.StatusInvalidArgument,
+			Msg:  fmt.Sprintf("Cannot handle data type `%s`, type=`%s`", dataType, typeString),
+		}
+	}
+	return field, nil
+}
+
+func parsePrecisionAndScale(name, typeString string) (int32, int32, error) {
+	typeString = strings.TrimSpace(typeString)
+	if len(typeString) == 0 {
+		return 0, 0, adbc.Error{
+			Code: adbc.StatusInvalidData,
+			Msg:  fmt.Sprintf("Cannot parse data type `%s` for field `%s`", typeString, name),
+		}
+	}
+
+	if typeString == "NUMERIC" {
+		return 38, 9, nil
+	} else if typeString == "BIGNUMERIC" {
+		return 76, 38, nil
+	}
+
+	values := strings.Split(strings.TrimRight(typeString[strings.Index(typeString, "(")+1:], ")"), ",")
+	if len(values) != 2 {
+		return 0, 0, adbc.Error{
+			Code: adbc.StatusInvalidData,
+			Msg:  fmt.Sprintf("Cannot parse data type `%s` for field `%s`", typeString, name),
+		}
+	}
+
+	precision, err := strconv.ParseInt(values[0], 10, 32)
+	if err != nil {
+		return 0, 0, adbc.Error{
+			Code: adbc.StatusInvalidData,
+			Msg:  fmt.Sprintf("Cannot parse precision `%s`: %s", values[0], err.Error()),
+		}
+	}
+
+	scale, err := strconv.ParseInt(values[1], 10, 32)
+	if err != nil {
+		return 0, 0, adbc.Error{
+			Code: adbc.StatusInvalidData,
+			Msg:  fmt.Sprintf("Cannot parse scale `%s`: %s", values[1], err.Error()),
+		}
+	}
```

Comment:
It might be easier to parse these out with a regexp instead and find submatches or use names